src/api/process.c

```diff
@@ -63,60 +84,205 @@ typedef enum {
+    node = node->next;
+    free(temp);
+  }
+}
+
+
+static bool kill_list_init(process_kill_list_t *list) {
+  list->mutex = SDL_CreateMutex();
+  list->has_work = SDL_CreateCond();
+  list->work_done = SDL_CreateCond();
+  list->head = list->tail = NULL;
+  list->stop = false;
+  if (!list->mutex || !list->has_work || !list->work_done) {
+    kill_list_free(list);
+    return false;
+  }
+  return true;
+}
+
+
+static void kill_list_push(process_kill_list_t *list, process_kill_t *task) {
+  if (!list) return;
+  task->next = NULL;
+  if (list->tail) {
+    list->tail->next = task;
+    list->tail = task;
+  } else {
+    list->head = list->tail = task;
+  }
+}
+
+
+static void kill_list_wait_all(process_kill_list_t *list) {
+  SDL_LockMutex(list->mutex);
+  // wait until list is empty
+  while (list->head)
+    SDL_CondWait(list->work_done, list->mutex);
+  // tell the worker to stop
+  list->stop = true;
+  SDL_CondSignal(list->has_work);
+  SDL_UnlockMutex(list->mutex);
+}
+
+
+static void process_handle_close(process_handle *handle) {
+#ifdef _WIN32
+  if (*handle) {
+    CloseHandle(*handle);
+    *handle = NULL;
+  }
+#endif
+  (void) 0;
+}
+
+
+static bool process_handle_is_running(process_handle handle, int *status) {
+#ifdef _WIN32
+  DWORD s;
+  if (GetExitCodeProcess(handle, &s) && s != STILL_ACTIVE) {
+    if (status != NULL)
+      *status = s;
+    return false;
+  }
+#else
+  int s;
+  if (waitpid(handle, &s, WNOHANG) != 0) {
+    if (status != NULL)
+      *status = WEXITSTATUS(s);
+    return false;
+  }
+#endif
+  return true;
+}
+
+
+static bool process_handle_signal(process_handle handle, signal_e sig) {
+#if _WIN32
+  switch(sig) {
+    case SIGNAL_TERM: return GenerateConsoleCtrlEvent(CTRL_BREAK_EVENT, GetProcessId(handle));
+    case SIGNAL_KILL: return TerminateProcess(handle, -1);
+    case SIGNAL_INTERRUPT: return DebugBreakProcess(handle);
+  }
+#else
+  switch (sig) {
+    case SIGNAL_TERM: return kill(-handle, SIGTERM) == 0; break;
+    case SIGNAL_KILL: return kill(-handle, SIGKILL) == 0; break;
+    case SIGNAL_INTERRUPT: return kill(-handle, SIGINT) == 0; break;
+  }
+#endif
+  return false;
+}
+
+
+static int kill_list_worker(void *ud) {
+  process_kill_list_t *list = (process_kill_list_t *) ud;
+  process_kill_t *current_task, *last_task;
+
+  while (true) {
+    SDL_LockMutex(list->mutex);
+
+    // wait until we have work to do
+    while (!list->head && !list->stop)
+      SDL_CondWaitTimeout(list->has_work, list->mutex, PROCESS_TERM_DELAY); // LOCK MUTEX
+
+    if (list->stop) break;
+
+    // filter the linked list and remove stale tasks
+    current_task = list->head;
+    last_task = NULL;
+    while (current_task) {
+      process_kill_t *temp;
+
+      // do not check for completion if the timeout has not expired
+      if ((SDL_GetTicks() - current_task->start_time) < PROCESS_TERM_DELAY) {
+        last_task = current_task;
+        current_task = current_task->next;
+        continue;
```

Comment:
If the head task hasn't expired, aren't all other tasks the same?
What I mean is that tasks are added by a single thread all with the same timeout, and are always added to the tail. So the first one that could possibly expire is the first one, right?

Also what do you think about using `SDL_Delay` with the time until the first task?
Basically:
1. Remove task from list
2. Check timeout
3. If timed out, goto 7
4. Unlock mutex
5. SDL_Delay for needed time
6. Lock mutex
7. Sigterm it, then add it back to the queue with new timeout, or sigkill it if out of retries
8. Goto 1

Reply:
I would assume that will wait for let's say 50ms per task.
In the meantime, you can have 10 tasks all completed in the 50ms wait period.

Instead of waiting once per task, it would be more efficient to wait once and sweep.
I'll agree with the first part, which I'll be fixing by breaking out of the loop.

Another method is process the tasks then enter an alertable sleep.
If this sleep completed, it will be stopped by the CondWait at the top.
Otherwise, it'll bypass the CondWait on the top and go straight to processing stuff.
This should make the thread "less spinny" when there's nothing.